### PR TITLE
Added ufw firewall application preset

### LINKED ISSUE #2435: "Uncomplicated FireWall" (UFW) Integration
It would be a nice to have an application preset for [ufw](https://help.ubuntu.com/community/UFW) in the `etc` folder of the releases (like for the startup scripts).
The file could look like this:
```ini
[syncthing]
title=Syncthing   
description=Sycthing file synchronisation
ports=22000/tcp|21027/udp
```
So (after copying it to `/etc/ufw/applications.d/`) opening the needed ports is as simple as
```
sudo ufw allow syncthing
```
The file could also be automatically added when installing with a package manager I guess.

### PR DESCRIPTION
**Pull request by Stefan-Code**
This closes #2435 
Also updated .gitignore to not exclude the newly added preset for ufw.

### CODE REVIEW

File: etc/firewall-ufw/syncthing [hunk 1]
```diff
@@ -0,0 +1,4 @@
+[syncthing]
+title=Syncthing   
+description=Sycthing file synchronisation
```
Comment (canton7) [inline etc/firewall-ufw/syncthing]: Typo 'sycthing'

Comment (Stefan-Code) [inline etc/firewall-ufw/syncthing]: Damn it. Thanks for the heads up, fixing it

**Comment by canton7** [comment]
:+1:

**Comment by Stefan-Code** [comment]
Does anyone know how to properly install this file with the debian package? When looking at other projects it seems like they move their preset to `/etc/ufw/applications.d/` regardless of whether ufw is installed or not. Is this proper behavior? Because then it would be as simple as adding the file [here]([https://github.com/syncthing/syncthing/blob/7351217489d708cccec2e52b6b061d1501e4aa8c/build.go#L301)

**Comment by KarlHegbloom** [comment]
Yes, that's right. The file does nothing, and is harmless, when the package
that reads it is not installed. So as long as it is not overwriting any
file from any other package, it's fine.

**Comment by KarlHegbloom** [comment]
You should also check and see if the ufw package should be given a
'force-reload' message in your postinst script. Use the package contents
search engine on the debian web to find other packages that install
conffiles for ufw and see what they do. Also look in Debian policy and
packaging manuals if ufw is a standard component now, to see if there's any
special thing for it you are supposed to do... actually,  if there is I
expect that lintian will catch it and tell you.

**Comment by Stefan-Code** [comment]
Is the exact build process you are using to build the `.deb` file documented somewhere?
When using
```
./build.sh deb
```
and
```
dpkg --build deb
```
I don't get any errors, however `lintian` doesn't really like the package:
```
E: syncthing: unstripped-binary-or-object usr/bin/syncthing
W: syncthing: hardening-no-relro usr/bin/syncthing
E: syncthing: debian-changelog-file-missing
E: syncthing: file-in-etc-not-marked-as-conffile etc/ufw/applications.d/syncthing
W: syncthing: unknown-control-file changelog
W: syncthing: unknown-control-file compat
E: syncthing: no-copyright-file
E: syncthing: description-contains-tabs
W: syncthing: no-section-field
W: syncthing: no-priority-field
E: syncthing: wrong-file-owner-uid-or-gid ..deb 1000/1000
W: syncthing: file-in-unusual-dir ..deb
E: syncthing: wrong-file-owner-uid-or-gid etc/ 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid etc/ufw/ 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid etc/ufw/applications.d/ 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid etc/ufw/applications.d/syncthing 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid lib/ 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid lib/systemd/ 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid lib/systemd/system/ 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid lib/systemd/system/syncthing@.service 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid usr/ 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid usr/bin/ 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid usr/bin/syncthing 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid usr/lib/ 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid usr/lib/systemd/ 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid usr/lib/systemd/user/ 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid usr/lib/systemd/user/syncthing.service 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid usr/share/ 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid usr/share/doc/ 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid usr/share/doc/syncthing/ 1000/1000
E: syncthing: bad-owner-for-doc-file usr/share/doc/syncthing/ sk/sk != root/root
E: syncthing: wrong-file-owner-uid-or-gid usr/share/doc/syncthing/AUTHORS.txt 1000/1000
E: syncthing: bad-owner-for-doc-file usr/share/doc/syncthing/AUTHORS.txt sk/sk != root/root
E: syncthing: wrong-file-owner-uid-or-gid usr/share/doc/syncthing/LICENSE.txt 1000/1000
E: syncthing: bad-owner-for-doc-file usr/share/doc/syncthing/LICENSE.txt sk/sk != root/root
W: syncthing: extra-license-file usr/share/doc/syncthing/LICENSE.txt
E: syncthing: wrong-file-owner-uid-or-gid usr/share/doc/syncthing/README.txt 1000/1000
E: syncthing: bad-owner-for-doc-file usr/share/doc/syncthing/README.txt sk/sk != root/root
E: syncthing: wrong-file-owner-uid-or-gid usr/share/man/ 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid usr/share/man/man1/ 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid usr/share/man/man1/syncthing.1 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid usr/share/man/man5/ 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid usr/share/man/man5/syncthing-config.5 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid usr/share/man/man5/syncthing-stignore.5 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid usr/share/man/man7/ 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid usr/share/man/man7/syncthing-device-ids.7 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid usr/share/man/man7/syncthing-event-api.7 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid usr/share/man/man7/syncthing-faq.7 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid usr/share/man/man7/syncthing-networking.7 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid usr/share/man/man7/syncthing-rest-api.7 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid usr/share/man/man7/syncthing-security.7 1000/1000
E: syncthing: wrong-file-owner-uid-or-gid usr/share/man/man7/syncthing-versioning.7 1000/1000
E: syncthing: manpage-not-compressed usr/share/man/man1/syncthing.1
E: syncthing: manpage-not-compressed usr/share/man/man5/syncthing-config.5
E: syncthing: manpage-not-compressed usr/share/man/man5/syncthing-stignore.5
E: syncthing: manpage-not-compressed usr/share/man/man7/syncthing-device-ids.7
E: syncthing: manpage-not-compressed usr/share/man/man7/syncthing-event-api.7
E: syncthing: manpage-not-compressed usr/share/man/man7/syncthing-faq.7
E: syncthing: manpage-not-compressed usr/share/man/man7/syncthing-networking.7
E: syncthing: manpage-not-compressed usr/share/man/man7/syncthing-rest-api.7
E: syncthing: manpage-not-compressed usr/share/man/man7/syncthing-security.7
E: syncthing: manpage-not-compressed usr/share/man/man7/syncthing-versioning.7
```
The `file-in-etc-not-marked-as-conffile` is caused by my modification.
Is this caused by my build process or something else?

**Comment by Stefan-Code** [comment]
Just opened a [seperate issue](#2441) regarding the `.deb` building.

**Comment by Stefan-Code** [comment]
Fixed the lintian error caused by my modifications. The deb building still needs a cleanup in general though IMHO.

**Comment by calmh** [comment]
LGTM. Can you squash to a single commit and reformat the commit message to adhere to the [guidelines](http://docs.syncthing.net/dev/contributing.html#commits)?

**Comment by calmh** [comment]
@st-jenkins add to whitelist

**Comment by Stefan-Code** [comment]
@calmh Allright, will do that tomorrow. Am I also responsible for adding myself to authors?

**Comment by Stefan-Code** [comment]
@calmh squashed and updated commit message

**Comment by calmh** [comment]
@st-jenkins retest this please

**Comment by calmh** [comment]
@st-jenkins retest this please

**Comment by calmh** [comment]
:+1: Thanks

**Comment by Stefan-Code** [comment]
No problem, glad to help at least a bit.

**Comment by calmh** [comment]
Reverted due to bug above, needs to be re-merged once we've repaired the upgrade system

**Comment by Stefan-Code** [comment]
Can't help on fixing this as I don't really know go, sorry.

**Comment by Stefan-Code** [comment]
@calmh Because #2446 is fixed by #2449 this could be re-merged I think.

**Comment by canton7** [comment]
Remember that #2449 needs to be released, *and* the vast majority of people need to be running it, before this can be re-merged.

Not everyone updates version-to-version. Some people leave their computers off for long periods, for example. If this is merged, anyone who is on 0.12.2 or earlier and tries to upgrade will fail.

**Comment by Stefan-Code** [comment]
Oh right. Didn't think this one through...

**Comment by calmh** [comment]
Yeah, sorry. We need to get most people off of v0.12.2 first. Or, we could call the file something other than `syncthing` in the meantime.

**Comment by Stefan-Code** [comment]
We could include it in the `etc` folder as e.g. `syncthing-firewall-preset.txt` but not include it in build.go and change the name while copying it to the deb build once most people are running v0.12.2.

**Comment by canton7** [comment]
Yeah, bugs in your upgrade system are a bitch :smile:

**Comment by calmh** [comment]
v0.10 not understanding how to upgrade to v0.11 was real fun too

**Comment by canton7** [comment]
I moved away from using github directly as the update server for partly this reason: I can mandate that all version A devices have to upgrade to B before upgrading to C. It is yet another bit of infrastructure though...

**Comment by calmh** [comment]
Yep. We could do that now if we wanted to though; it would "just" take implementing the server side, and changing the default URL. The new server side could decide what versions to announce by an X-Syncthing-Version header sent by the client, and otherwise behave exactly like the GitHup API.

**Comment by meonkeys** [comment]
Just wanted to say thanks for the Syncthing ufw config file... super handy, and I learned a bit about ufw from it! Are folks finally mostly at versions > v0.12.2?

**Comment by calmh** [comment]
Yes, I think there's a negligible amount users left on that. We should add this to the distribution archives now.

**Comment by Stefan-Code** [comment]
@calmh Should I open another pull request for this again then?

**Pull request merged**